Timing data in Sampler (#1909)
Qiskit/documentation: docs/guides/monitor-job.mdx

|                               |                                                                                                                                                                                                              |
| ----------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| job.result()                  | Review job results immediately after the job completes. Job results are available after the job completes. Therefore, job.result() is a blocking call until the job completes.                               |

> | job.job\_id()                 | Return the ID that uniquely identifieshttps://ibm-quantumcomputing.slack.com/archives/CSV6D7KKL/p1725881005460219 that job. Retrieving the job results at a later time requires the job ID. Therefore, it is recommended that you save the IDs of jobs you might want to retrieve later. |

Reviewer: ```suggestion
| job.job\_id()                 | Return the ID that uniquely identifies that job. Retrieving the job results at a later time requires the job ID. Therefore, it is recommended that you save the IDs of jobs you might want to retrieve later. |
```

In Qiskit, use `job.cancel()` to cancel a job.

<span id="execution-spans"></span>
## View Sampler execution spans

> The results of `SamplerV2` jobs executed in Qiskit Runtime contain execution timing information in their metadata.

Reviewer 2: Out of curiosity, is there some kind of magic that goes through these markdowns and convert class names to API links?

Author: No; I need to add links if we want them.

Author: I added a link to the first reference.

This timing information can be used to place upper and lower timestamp bounds on when particular shots were executed on the QPU.
Shots are grouped into `~.ExecutionSpan`s, each of which indicates a start time, a stop time, and a specification of which shots were collected in the span.

> An execution span specifies which data was executed during its window by providing an `~.ExecutionSpan.mask` method. This method, given any PUB index, returns a boolean mask that is ``True`` for all shots executing during its window. PUBs are indexed by the order in which they were given to the Sampler run call. If, for example, a PUB has shape `(2,3)` and was run with four shots, then the mask's shape is `(2, 3, 4)`.

Reviewer 2: ```suggestion
An execution span specifies which data was executed during its window by providing an `~.ExecutionSpan.mask` method. This method, given any PUB index, returns a boolean mask that is ``True`` for all shots executing during its window. PUBs are indexed by the order in which they were given to the Sampler run call. If, for example, a PUB has shape `(2, 3)` and was run with four shots, then the mask's shape is `(2, 3, 4)`.
```

> An execution span specifies which data was executed during its window by providing an `~.ExecutionSpan.mask` method. This method, given any PUB index, returns a boolean mask that is ``True`` for all shots executing during its window. PUBs are indexed by the order in which they were given to the Sampler run call. If, for example, a PUB has shape `(2, 3)` and was run with four shots, then the mask's shape is `(2, 3, 4)`.

Reviewer 2: ```suggestion
An execution span specifies which data was executed during its window by providing an `~.ExecutionSpan.mask` method. This method, given any PUB index, returns a boolean mask that is ``True`` for all shots executed during its window. PUBs are indexed by the order in which they were given to the Sampler run call. If, for example, a PUB has shape `(2, 3)` and was run with four shots, then the mask's shape is `(2, 3, 4)`.
```

```python
# Get the mask of the 1st PUB for the 0th span.
mask = spans[0].mask(1)

# Decide whether the 0th shot of parameter set (1, 2) occurred in this span.
in_this_span = mask[1, 2, 0]

# Create a new bit array containing only the PUB-1 data collected during this span.
bits = pub_results[1].data.meas

> filtered_data = BitArray(bits.array(mask), bits.num_bits)

Reviewer 2: ```suggestion
filtered_data = BitArray(bits.array[mask], bits.num_bits)
```

```

To view execution span information, review the metadata of the result returned by `SamplerV2`, which comes in the form of an `~ExecutionSpans` object. This object is a list-like container containing subclass instances of `ExecutionSpan`s such as `SliceSpan`:

```python
pub_results = job.result()
spans = pub_results.metadata["execution"]["execution_spans"]
for span in spans:

>      print(span)

Reviewer 2: Is there any chance we could show an example of what the output looks like? Here,

```bash
SliceSpan(<start='2024-09-09 11:48:21', stop='2024-09-09 11:48:22', size=1280>)
SliceSpan(<start='2024-09-09 11:48:21', stop='2024-09-09 11:48:25', size=1280>)
SliceSpan(<start='2024-09-09 11:48:21', stop='2024-09-09 11:48:24', size=1280>)
SliceSpan(<start='2024-09-09 11:48:21', stop='2024-09-09 11:48:27', size=1280>)
SliceSpan(<start='2024-09-09 11:48:21', stop='2024-09-09 11:48:26', size=1280>)
...
```

```

Execution spans can be filtered to include information pertaining to specific PUBs, selected by their indices:

```python
# take the subset of spans that reference data in PUBs 0 or 2
spans.filter_by_pub([0, 2])
```

View global information about the collection of execution spans:

```python
print("Number of execution spans:", len(spans))
print("  Start of the first span:", spans.start)
print("     End of the last span:", spans.stop)
print("       Total duration (s):", spans.duration)
```

Extract and inspect a particular span:

```python
spans.sort()
print(" Start of first span:", spans[0].start)
print("   End of first span:", spans[0].stop)
print("#shots in first span:", spans[0].size)
```

> <Admonition>

Reviewer 2: This is more about the theme than anything, but Admonition seems like rather a subtle change over the regular paragraph style, and not very admonishing :P

![image](https://github.com/user-attachments/assets/4023226f-389a-42ac-9b40-1f9a0bf05ebf)

Author: Oh, I didn't put a type.  That will help!

In Qiskit, use `job.cancel()` to cancel a job.

<span id="execution-spans"></span>
## View Sampler execution spans

The results of [`SamplerV2`](/api/qiskit-ibm-runtime/qiskit_ibm_runtime.SamplerV2) jobs executed in Qiskit Runtime contain execution timing information in their metadata.
This timing information can be used to place upper and lower timestamp bounds on when particular shots were executed on the QPU.
Shots are grouped into `~.ExecutionSpan`s, each of which indicates a start time, a stop time, and a specification of which shots were collected in the span.

> An execution span specifies which data was executed during its window by providing an `~.ExecutionSpan.mask` method. This method, given any PUB index, returns a boolean mask that is ``True`` for all shots executed during its window. PUBs are indexed by the order in which they were given to the Sampler run call. If, for example, a PUB has shape `(2, 3)` and was run with four shots, then the mask's shape is `(2, 3, 4)`.

Reviewer: ```suggestion
An execution span specifies which data was executed during its window by providing an `~.ExecutionSpan.mask` method. This method, given any PUB index, returns a Boolean mask that is ``True`` for all shots executed during its window. PUBs are indexed by the order in which they were given to the Sampler run call. If, for example, a PUB has shape `(2, 3)` and was run with four shots, then the mask's shape is `(2, 3, 4)`.
```

print("#shots in first span:", spans[0].size)
```

> <Admonition type ="note">

Reviewer 3: ```suggestion
<Admonition type="note">
```
